### DISCUSSION TITLE
How to Change Author Name and Email of Commits?

### DISCUSSION THREAD

**Discussion by a-a-ron** [community/community#22185] [New to GitHub]
<p>How can i change the author name and email settings globally for future commits? Is there also a way to just do this per repository? Can i also change this information on past commits?  </p>

**Comment by lee-dohm** [comment, marked as answer]
<p>Great question, thanks for asking!</p>
<p>These are all pretty common questions that we have help pages for:</p>
<ul>
<li>
<a href="https://help.github.com/articles/setting-your-commit-email-address-in-git/" rel="nofollow noopener">Setting user email address</a> either globally or locally to a single repo (setting the author name is the same steps but putting user.name where user.email is)</li>
<li>
<a href="https://help.github.com/articles/changing-author-info/" rel="nofollow noopener">Changing historic author info</a> in a repository</li>
</ul>
<p>I hope that helps :grinning:</p>

**Reply by tertek** [reply]
@lee-dohm 
Unfortunately, help page about "Changing historic author info" is not available anymore. Is there any updated resource?

**Reply by airtower-luna** [reply]
See: [Git Tools - Rewriting History](https://git-scm.com/book/en/v2/Git-Tools-Rewriting-History)

**Comment by a-a-ron** [comment]
<p>Just the information i was looking for. Thanks!</p>

**Comment by nadiajoyce** [comment]
<p><a class="mention" href="/teazateazza">@teazateazza</a> It looks like you were maybe trying to respond here, but only the quote of <a class="mention" href="/lee-dohm">@lee-dohm</a>’s comment came through. You might want to edit your post to make sure your whole message is shared. Thanks!</p>

**Comment by ofiliz** [comment]
<p>Related question: What happens when we lose access to an email address (for example, when changing employers) that our previous commits were associated with. It looks like all the commit and contribution history is lost. Years worth of history. This is super inconvenient.</p>
<p>The second option you listed below doesn’t work since most authors don’t own the repos, and even if they did, rewriting git history is not a good practice.</p>
<p>GitHub knows that all these commits were made by the author, even if that author does not get to keep his/her email address forever. So ideally GitHub should be able to still associate those commits to the correct author.</p>
<p>Is there anyway to recover history? Thanks for your help.</p>

**Comment by lee-dohm** [comment]
<p>You don’t have to remove an email address from the list of email addresses in your GitHub account simply because you’ve lost access to it. You can leave it associated to your account to maintain a connection with your old history. You should make certain that you always have access to <em>at least</em> one email address associated with your GitHub account for password resets, email notifications and the like. But you don’t have to remove an email address just because it isn’t current.</p>
<p>The way the system currently works, associating commits by email address, is the best balance between multiple needs. There are also times that people want to divorce themselves from work they did in the past and removing the email address from the list of addresses that apply is an easy way to do that. If we used some other way of tracking contributions, removing unwanted items from your history would be a much more complex affair.</p>

**Comment by BobMarq** [comment]
<p>I applied for an account with the username BobMarq and then accidentally entered a mispelled e-mail address</p>

**Comment by jojomatik** [comment]
<p>The second link isn’t working for me.</p>
<p>Has it been removed on purpose?</p>
<p>If not, could you please update the link or post a new working one?</p>

**Comment by ahmed2m** [comment]
<p>This is the most recent version I can find working<br>
</p><aside class="onebox allowlistedgeneric">
  <header class="source">
      
      <a href="https://docs.github.com/en/enterprise/2.18/user/github/using-git/changing-author-info" target="_blank" rel="noopener nofollow ugc">docs.github.com</a>
  </header>
  <article class="onebox-body">
    <img src="" class="thumbnail" width="" height="">

<h3><a href="https://docs.github.com/en/enterprise/2.18/user/github/using-git/changing-author-info" target="_blank" rel="noopener nofollow ugc">Changing author info - GitHub Docs</a></h3>



  </article>
  <div class="onebox-metadata">
    
    
  </div>
  <div style="clear: both"></div>
</aside>

**Comment by vbresan** [comment]
<p>This did not work for me. After I executed the script I got the following:</p>
<pre><code class="lang-auto">WARNING: git-filter-branch has a glut of gotchas generating mangled history
         rewrites.  Hit Ctrl-C before proceeding to abort, then use an
         alternative filtering tool such as 'git filter-repo'
         (https://github.com/newren/git-filter-repo/) instead.  See the
         filter-branch manual page for more details; to squelch this warning,
         set FILTER_BRANCH_SQUELCH_WARNING=1.
Proceeding with filter-branch...

Rewrite 9e53679d...
...
WARNING: Ref 'refs/heads/master' is unchanged
WARNING: Ref 'refs/heads/wiki' is unchanged
</code></pre>
<p>Then I executed:<br>
<code>git push --force --tags origin 'refs/heads/*'</code></p>
<p>And got the following message:<br>
<code>Everything up-to-date</code></p>
<p>When I reloaded the repository on Github, no changes took effect.</p>
<p>Anyone could suggest what has to be done? Thanks!</p>

**Comment by vbresan** [comment]
<p>Apparently, in my case email address was used as either <code>GIT_COMMITTER_NAME</code> or <code>GIT_AUTHOR_NAME</code> and that’s why the script was not working. That probably happened when I was importing the source code from google code a while ago.</p>
<p>Extending <code>if</code> statement conditions to:<br>
<code>if [ "$GIT_COMMITTER_EMAIL" = "$OLD_EMAIL" ] || [ "$GIT_COMMITTER_NAME" = "$OLD_EMAIL" ]</code></p>
<p>and</p>
<p><code>if [ "$GIT_AUTHOR_EMAIL" = "$OLD_EMAIL" ] || [ "$GIT_AUTHOR_NAME" = "$OLD_EMAIL" ]</code></p>
<p>made it work.</p>

**Comment by kavonszadkowski** [comment]
<p>I was not aware of this and recently cleaned up my account information (as you do with many accounts when your email addresses change) and just realized that I thus lost 5 years worth of work associated with those old addresses.<br>
Is there any way to undo deleting old email addresses (they don’t exist any more, so simply re-adding wouldn’t work)? I’m still the admin of the repositories in question, I just don’t want to rewrite history by changing the commits.</p>
<p>Maybe a straightforward solution to this whole issue would be to add an option in the settings for users to choose whether to list contributions by email address or account. It doesn’t seem likely to me that the majority of users wish to disassociate themselves from previous work, so association-by-account seems the more reasonable default. In contrast kicking past contributions off your account should be an active choice and not something that simply happens as a by-product of basic account maintenance.</p>

**Comment by gnomefin** [comment]
I think the above liink is not valid anymore. This is my way: Please first clone your desired repository
```
# Check first your email commit
git log --pretty="format:%ae"

#please find your old email address that associated with commit contribution.
#please find your new email address based on your primary associated email, or using the generic github noreply email. you can actually find it `.*@users.noreply.github.com`

# execute
#!/bin/sh
git filter-branch --env-filter '
OLD_EMAIL="xxx@gmail.com"
CORRECT_NAME="Alfian Firmansyah"
CORRECT_EMAIL="accountnumber+yourusername@users.noreply.github.com"
if [ "$GIT_COMMITTER_EMAIL" = "$OLD_EMAIL" ]
then
    export GIT_COMMITTER_NAME="$CORRECT_NAME"
    export GIT_COMMITTER_EMAIL="$CORRECT_EMAIL"
fi
if [ "$GIT_AUTHOR_EMAIL" = "$OLD_EMAIL" ]
then
    export GIT_AUTHOR_NAME="$CORRECT_NAME"
    export GIT_AUTHOR_EMAIL="$CORRECT_EMAIL"
fi
' --tag-name-filter cat -- --branches --tags

# push
git push --force --tags origin 'refs/heads/*'

# And voila, all the contributors are moved to your new email address.
```

**Comment by nanisadw3** [comment]
### 1. Fix only the most recent commit:
```bash
git commit --amend --author="Your Name <your-email@example.com>" --no-edit
```
### 2. Fix multiple past commits (Recommended tool: `git-filter-repo`):
The modern and safe standard recommended by the Git team is `git-filter-repo`:
```bash
git filter-repo --email-callback '
return email.replace(b"old-email@example.com", b"new-email@example.com")
'
```
Alternatively, to update all authors in the current repo using the configured git user:
```bash
git rebase -r --root --exec "git commit --amend --reset-author --no-edit"
```
*Note:* If you've already pushed these commits, you will need to push with `git push --force-with-lease`.
If this was helpful, please feel free to mark this as the accepted answer!
